File: Problems 0100-0199/135. Candy/solution.py
class Solution:
    def candy(self, ratings: List[int]) -> int:
        numChildren = len(ratings)
        candies = [1] * numChildren # every child gets 1 candy
        for i in range(1, numChildren):
            if ratings[i] > ratings[i-1]:
                candies[i] = candies[i-1] + 1
        
        for i in range(numChildren-2, -1, -1):
            if ratings[i] > ratings[i+1]:
                candies[i] = max(candies[i], candies[i+1] + 1)

        return sum(candies)

File: Problems 0100-0199/135. Candy/solution.cpp
class Solution {
    public:
        int candy(vector<int>& ratings) {
            // peaks and valleys
            // whenever we reach a valley, it should always be 1
            // and we should always climb up by one until we reach the peak
            // the peak needs to be readjusted if the slope down needs it
            // to be more
            int n = ratings.size();
            int res = n; // minimum 1 candy per child
            int i = 1;
            while (i < n) {
                if (ratings[i] == ratings[i-1]) {
                    i++;
                    continue;
                }
    
                int peak = 0;
                while (i < n && ratings[i] > ratings[i-1]) {
                    peak++;
                    res += peak;
                    i++;
    
                    if (i == n) return res;
                }
    
                int valley = 0;
                while (i < n && ratings[i] < ratings[i-1]) {
                    valley++;
                    res += valley;
                    i++;
                }
    
                res -= min(peak, valley);
    
            }
    
            return res;
        }
    };